### PR TITLE
Integration test

### PR DESCRIPTION
**Pull request by skyfrench**
Integration testing.
Implemented for Excalibur in [#36](https://github.com/dls-controls/excalibur-detector/pull/36)

### CODE REVIEW

File: common/include/DummyUDPDefinitions.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  } FrameHeader;
+  
+  inline const std::size_t max_frame_size(void)
+  {
+    std::size_t max_frame_size = sizeof(FrameHeader) + (max_packet_size * max_packets);
+    return max_frame_size;
+  }
+}
+
+#endif /* INCLUDE_DUMMYUDP_DEFINITIONS_H_ */
```
Comment (MattTaylorDLS) [inline common/include/DummyUDPDefinitions.h]: Github complaining about no new line at end of file

File: frameSimulator/include/FrameSimulatorOptionsDummyUDP.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    static const int default_image_width = 1024;
+    static const int default_image_height = 1024;
+    static const int default_pkt_len = 8000;
+
+    static const FrameSimulatorOption<int> opt_image_width("width", "Simulated image width");
+    static const FrameSimulatorOption<int> opt_image_height("height", "Simulated image height");
+    static const FrameSimulatorOption<int> opt_packet_len("packet-len", "Length of simulated packets in bytes");
+
+}
+#endif // FRAMESIMULATOR_FRAMESIMULATOROPTIONSDUMMYUSP_H
```
Comment (MattTaylorDLS) [inline frameSimulator/include/FrameSimulatorOptionsDummyUDP.h]: Github complaining about no newline at end of file

File: frameSimulator/src/FrameSimulatorPluginUDP.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        } else {
+            create_frames(replay_numframes.get());
+        }
+
+        return true;
+
+    }
+
+    /** Prepare frame(s) for replay by processing the pcap file
+     * /param[in] pcap header
```
Comment (MattTaylorDLS) [inline frameSimulator/src/FrameSimulatorPluginUDP.cpp]: "pcap" doesn't match parameter name

File: test/integrationTest/config/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+install(FILES dummyUDP.json DESTINATION test_config)
+install(FILES dummyUDP-fp.json DESTINATION test_config)
+install(FILES dummyUDP-fr.json DESTINATION test_config)
+install(FILES testUDP.json DESTINATION test_config)
```
Comment (MattTaylorDLS) [inline test/integrationTest/config/CMakeLists.txt]: Github complaining about no newline at end of file

File: test/integrationTest/src/ControlUtility.cpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+namespace FrameSimulatorTest {
+
+    /** ControlUtility constructor
+     *
+     * @param ptree - boost property_tree; parsed from initial ini file
+     * @param positional_arg - (possible) positional argument
+     * @param process_entry - property tree entry accessor for process path
+     * @param process_args_entry - property tree entry accessor for process arguments
+     * @param process_pid - pid of process
```
Comment (MattTaylorDLS) [inline test/integrationTest/src/ControlUtility.cpp]: missing parameter comments for socket_entry and kill_entry, and no process_id in arguments

File: test/integrationTest/src/ControlUtility.cpp [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+          command += command_args_[s] + " ";
+        }
+
+      int status = std::system((command).c_str());
+      exit_code_ = WEXITSTATUS(status);
+
+    }
+
+    /** Send configuration message
+     *
```
Comment (MattTaylorDLS) [inline test/integrationTest/src/ControlUtility.cpp]: needs parameter comment

File: test/integrationTest/src/FrameTestApp.cpp [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+using namespace FrameSimulatorTest;
+
+namespace po = boost::program_options;
+
+#include "logging.h"
+#include "DebugLevelLogger.h"
+
+/** Check that str contains suffix
+ * /param[in] str - string to test
+ * /param[in] str - test suffix
```
Comment (MattTaylorDLS) [inline test/integrationTest/src/FrameTestApp.cpp]: str doesn't match suffix parameter name

File: test/integrationTest/src/FrameTestApp.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      std::string config_file = vm["json"].as<std::string>();
+      LOG4CXX_DEBUG(logger, "Reading config file " << config_file);
+      boost::property_tree::json_parser::read_json(config_file, ptree);
+    } else {
+      LOG4CXX_ERROR(logger, "No configuration file specified. Exiting.");
+      exit(1);
+    }
+
+  }
+
```
Comment (MattTaylorDLS) [inline test/integrationTest/src/FrameTestApp.cpp]: blank line between } and catch

File: test/integrationTest/src/FrameTestApp.cpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    BOOST_FOREACH(boost::property_tree::ptree::value_type &vc, pt.get_child("Main")) {
+      std::string process = vc.first;
+      LOG4CXX_INFO(logger, "Process to launch: " + process);
+      std::string command_entry = "Main." + process + "." + "command";
+      std::string pos_args = pt.get<std::string>("Main." + process + "." + "pos-args");
+      std::string socket_entry = "Main." + process + "." + "socket";
+      std::string kill_entry = "Main." + process + "." + "kill";
+      int sleeptime = pt.get<int>("Main." + process + "." + "sleep");
+      ControlUtility* control = new ControlUtility(pt, pos_args, command_entry, process, socket_entry, kill_entry, logger);
+            utilities.push_back(control);
```
Comment (MattTaylorDLS) [inline test/integrationTest/src/FrameTestApp.cpp]: formatting seems off

File: test/integrationTest/src/PropertyTreeUtility.cpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+                updated_file.open(json_file.c_str(), std::ios::trunc);
+                updated_file << new_json;
+                updated_file.close();
+              }
+
+            }
+    }
+
+
+}
```
Comment (MattTaylorDLS) [inline test/integrationTest/src/PropertyTreeUtility.cpp]: Github complaining about no newline at end of file

File: frameReceiver/src/DummyUDPFrameDecoder.cpp [hunk 11]
```diff
@@ -50,36 +75,449 @@ std::string DummyUDPFrameDecoder::get_version_long()
...
+}
+
+//! Calculate and return an elapsed time in milliseconds.
+//!
+//! This method calculates and returns an elapsed time in milliseconds based on the start and
+//! end timespec structs passed as arguments.
+//!
+//! \param[in] start - start time in timespec struct format
+//! \param[in] end - end time in timespec struct format
+//! \return eclapsed time between start and end in milliseconds
```
Comment (MattTaylorDLS) [inline frameReceiver/src/DummyUDPFrameDecoder.cpp]: typo - eclapsed

**Comment by skyfrench** [comment]
> Only minor formatting comments.

All resolved.

> One question though: Do we have/plan to have a Dummy ZMQ plugin an associated tests?

I propose we merge this without - and open a new pull request when a Dummy ZMQ plugin is written? Then we can close all the related merge requests on the Excalibur etc.

**Comment by skyfrench** [comment]
> 1. Minor point, but the images look like they are square.  It would be good to run with different width and height to make sure we aren't flipping them somewhere.

It's now rectangular ...

> 2. The first run appeared to suggest that the replay/simulator had sent all packets without packet loss but all frames were reported as containing missing packets.  Is this expected?  Are we sending the packets through all at once?

Packets are getting lost. Any tips on how to stop this?

**Pull request merged**